File: skills/vector-index/scripts/guards.py
#!/usr/bin/env python3
"""
guards — capability & environment guards (C9).

Generalized from the codebase-RAG document, which gated its whole index/MCP
surface behind an environment check at three layers (route, server, client). The
plugin is read-mostly, but `ingest` / `reindex` / `create_project` / `add_source`
mutate state and read arbitrary filesystem paths, so they declare an enablement
guard enforced *before any logic runs*:

    VINDEX_READONLY=1            block all mutating operations
    VINDEX_ALLOW_ROOTS=a:b:c     only ingest/create from under these roots

Stdlib only; reads the environment live so a host can flip the policy per process.
Guards return an error dict (so MCP tools surface it) or None to proceed.
"""

from __future__ import annotations

import os
from pathlib import Path

_TRUE = {"1", "true", "yes", "on"}


def readonly() -> bool:
    return os.environ.get("VINDEX_READONLY", "").strip().lower() in _TRUE


def deny_if_readonly(tool: str) -> dict | None:
    if readonly():
        return {"error": f"{tool} blocked: VINDEX_READONLY is set (read-only mode)"}
    return None


def allowed_roots() -> list[Path]:
    raw = os.environ.get("VINDEX_ALLOW_ROOTS", "")
    out: list[Path] = []
    for part in raw.split(os.pathsep):
        part = part.strip()
        if part:
            try:
                out.append(Path(os.path.expanduser(part)).resolve())
            except Exception:
                pass
    return out


def path_allowed(path: str | os.PathLike) -> bool:
    roots = allowed_roots()
    if not roots:
        return True  # no allow-list configured -> unrestricted
    try:
        p = Path(os.path.expanduser(str(path))).resolve()
    except Exception:
        return False
    return any(p == r or r in p.parents for r in roots)


def deny_if_path_blocked(tool: str, path: str | os.PathLike | None) -> dict | None:
    if path and not path_allowed(path):
        return {"error": f"{tool} blocked: {path} is outside VINDEX_ALLOW_ROOTS"}
    return None

File: src/guards.ts
/**
 * guards — capability & environment guards (C9).
 *
 * The plugin is read-mostly, but ingest / reindex / create_project / add_source
 * mutate state and read arbitrary filesystem paths, so they assert an enablement
 * guard *before any logic runs*:
 *
 *   VINDEX_READONLY=1          block all mutating operations
 *   VINDEX_ALLOW_ROOTS=a:b:c   only ingest/create from under these roots
 *
 * Policy is read from config.ts (resolved once at process start).
 */
import { homedir } from 'node:os'
import { isAbsolute, resolve } from 'node:path'
import { READONLY, ALLOW_ROOTS } from './config.ts'

/** Expand a leading `~` and resolve to an absolute, normalized path. */
function resolvePath (p: string): string {
  let s = p
  if (s === '~')
    s = homedir()
  else if (s.startsWith('~/'))
    s = homedir() + s.slice(1)
  return isAbsolute(s) ? resolve(s) : resolve(process.cwd(), s)
}

const ROOTS = ALLOW_ROOTS.map(resolvePath)

/** True when the process is in read-only mode. */
export function readonly (): boolean {
  return READONLY
}

/** Throw if mutations are disabled. `action` names the blocked operation. */
export function assertWritable (action = 'operation'): void {
  if (READONLY)
    throw new Error(`${action} blocked: VINDEX_READONLY is set (read-only mode)`)
}

/** True if `p` lies under some allow-root; unrestricted when none are configured. */
export function pathAllowed (p: string): boolean {
  if (ROOTS.length === 0)
    return true

  const target = resolvePath(p)
  return ROOTS.some(
    root => target === root || target.startsWith(root.endsWith('/') ? root : root + '/'),
  )
}

/** Throw if `p` is outside every ALLOW_ROOTS entry (no-op when list is empty). */
export function assertAllowedRoot (p: string): void {
  if (!pathAllowed(p))
    throw new Error(`${p} is outside VINDEX_ALLOW_ROOTS`)
}
